src/word_search_generator/types.py

```diff
@@ -1,19 +1,144 @@
-from typing import Optional, TypedDict
+from enum import Enum, unique
+from typing import NamedTuple, Optional, TypedDict
+
+
+@unique
+class Direction(Enum):
+    """
+    If you want custom directions, like `"skipE": (0, 2)`, this is the
+    place to monkey-patch them in.
+
+    Tuples are listed in (∂row, ∂col) pairs, presumably b/c that makes
+    it easier to use with the Puzzle = list[list[chr]] format
+    """
+
+    # is there a better way to specify typing here?
+    # without hints here, the linter gets upset with my definitions of r/c_move
+    N: tuple[int, int] = (-1, 0)
+    NE: tuple[int, int] = (-1, 1)
+    E: tuple[int, int] = (0, 1)
+    SE: tuple[int, int] = (1, 1)
+    S: tuple[int, int] = (1, 0)
+    SW: tuple[int, int] = (1, -1)
+    W: tuple[int, int] = (0, -1)
+    NW: tuple[int, int] = (-1, -1)
+
+    @property
+    def r_move(self) -> int:
+        return self.value[0]
+
+    @property
+    def c_move(self) -> int:
+        return self.value[1]
+
+
+class Position(NamedTuple):
+    row: int
+    column: int
 
 
 class KeyInfo(TypedDict):
-    start: tuple[int, int]
-    direction: str
+    start: Optional[Position]
+    direction: Optional[Direction]
+    secret: bool
 
 
 class KeyInfoJson(TypedDict):
-    start_row: int
-    start_col: int
-    direction: str
+    start_row: Optional[int]
+    start_col: Optional[int]
+    direction: Optional[str]
+    secret: bool
+
+
+class Word:
+    """This class represents a Word within a WordSearch puzzle."""
+
+    def __init__(
+        self,
+        text: str,
+        secret: bool = False,
+    ):
+        self.text = text.upper()
+        self._start_row: Optional[int] = None
+        self._start_column: Optional[int] = None
+        self.direction: Optional[Direction] = None
+        self.secret = secret
+
+    @property
+    def start_row(self) -> int | None:
+        if isinstance(self._start_row, int):
```

Comment:
For both this and `start_column`, do we need the `if isinstance(self._start, int)` check?

Without such a check, if the internal value is None, then it returns None; if it's an int, then it returns its numeric value.  Do we expect someone to set one of the private `_start` attributes to be a `set` or something?

Reply:
I was in type hinting hell and just trying to get mypy to quit bugging me about return type for the other properties. I'm sure I can rework or I just was over dealing with type hints at that point, to be honest. I'll check it out today.

Reply:
I certainly empathize with adjusting the otherwise-correct code to satisfy the type checker.

I'll give the rest of the changes a closer look this evening when I can pour over them in PyCharm instead of GH diffs.

Reply:
@duck57, I've got the typing cleaned up for this in the last commit (02bc02b). I did have to change the typing in the `generate.test_a_fit()` and `generate.find_a_fit()` functions from `Position` to just `tuple(int, int)` to keep mypy happy. Seems like a reasonable compromise, as it would require a bunch of other "useless/no effect on code function" checks to make the `Position` type work there.